main/src/ca/uwaterloo/flix/language/phase/Instances.scala

```diff
@@ -43,13 +43,33 @@ object Instances extends Phase[TypedAst.Root, TypedAst.Root] {
     */
   private def visitInstances(root: TypedAst.Root)(implicit flix: Flix): Validation[Unit, InstanceError] = {
 
+    // MATT test
+    /**
+      * Checks that an instance is not an orphan.
+      * It is declared in either:
+      * * The class's companion namespace.
+      * * The same namespace as its type.
+      */
+    def checkOrphan(inst: TypedAst.Instance): Validation[Unit, InstanceError] = inst match {
+      case TypedAst.Instance(_, _, classSym, tpe, _, _, ns, loc) => tpe.typeConstructor match {
+        // Case 1: Enum type in the same namespace as the instance: not an orphan
+        case Some(TypeConstructor.Enum(enumSym, _)) if enumSym.namespace == ns.idents.map(_.name) => ().toSuccess
+        // Case 1.5: Array type in the top-level namespace: not an orphan
+        case Some(TypeConstructor.Array) if  ns.isRoot => ().toSuccess // TODO hack?
+        // Case 2: Any type in the class companion namespace: not an orphan
+        case _ if (classSym.namespace :+ classSym.name) == ns.idents.map(_.name) => ().toSuccess
+        // Case 3: Any type outside the class companion namespace and enum declaration namespace: orphan
+        case _ => InstanceError.OrphanInstance(tpe, classSym, loc).toFailure
```

Comment:
Wait.. there are many type constructors... Arrow.. Channel... Int32 etc. What should we do?

Either:

- Lenient: It is only for enums that the orphan rule applies.
- Strict: You can only define instances next to the TC. Except for enums where you are also allowed to define them next to their declaration.

Reply:
If we go with lenient, it seems like an arbitrary restriction just to put it on enums.

I think strict makes more sense. Are there any issues with it?

Reply:
Let us do strict. Does the code need to change?

Reply:
I'll need to move cases out of Array.flix, but otherwise the logic is the same.

Reply:
I wasn't able to move the `Array` instance since its implementation uses `Array` library functions, which are not in the core library.

Reply:
Can you open a ticket?